mm2src/mm2_main/src/lp_swap/taker_swap.rs

```diff
@@ -1956,11 +2051,18 @@ impl TakerSwap {
             taker_coin,
             data.lock_duration,
             data.p2p_privkey.map(SerializableSecp256k1Keypair::into_inner),
+            #[cfg(any(test, feature = "run-docker-tests"))]
+            fail_at,
         );
         let command = saved.events.last().unwrap().get_command();
-        for saved_event in saved.events {
-            swap.apply_event(saved_event.event);
+        for saved_event in &saved.events {
+            swap.apply_event(saved_event.event.clone());
+        }
+
+        if check_watcher_payments(&swap, &ctx, saved).await? {
```

Comment:
Can this return `Result<(Self, Option<TakerSwapCommand>), String>` instead? I also think this function should execute code based on commands returned from here https://github.com/KomodoPlatform/komodo-defi-framework/blob/b0ffd0d92da386fd49155d5df69506d1b81c321a/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L2057 similar to how `handle_command` works https://github.com/KomodoPlatform/komodo-defi-framework/blob/b0ffd0d92da386fd49155d5df69506d1b81c321a/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L897 for instance if the next command was `TakerSwapCommand::SpendMakerPayment` we wouldn't need to check if taker payment is spent or not since it's a given https://github.com/KomodoPlatform/komodo-defi-framework/blob/b0ffd0d92da386fd49155d5df69506d1b81c321a/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L2435
What do you think about that @caglaryucekaya?

Reply:
I handled not checking for taker payment spend or refund if the `TakerPaymentSpent` event exists, but I didn't use `get_command` for it, I only check if there is `TakerPaymentSpent` event in the saved swap. I don't think using the commands in this function is necessary because the action to be taken doesn't depend on how the swap exactly failed or what the last event was, but rather if a watcher has spent the maker payment or refunded the taker payment. For example, it might be the case that the taker has failed to spend the maker payment and went offline while trying to refund the taker payment. Meanwhile, a watcher can successfully spend the maker payment and complete the swap successfully. In this case, the taker should also save the swap as successful by adding the event `"MakerPaymentSpentByWatcher"`.

About `check_watcher_payments` function returning `Result<(Self, Option<TakerSwapCommand>), String>`, what the function does is checking for maker payment spend and taker payment refund and modifying the saved events if necessary, or do nothing.

https://github.com/KomodoPlatform/komodo-defi-framework/blob/47c2257796cecfc1cec09d844230f6709d2568fc/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L2076-L2085

If maker payment spend or taker payment refund is found and the saved events are modified, it returns true, and `load_from_saved ` function returns with the `TakerSwapCommand::Finish` command to bypass the kickstart. If no payments are found, `check_watcher_payments` does nothing and the `load_from_saved ` returns the next command and kick starts the swap as usual. I can either leave the function as it is and return a boolean, which then makes the `load_from_saved ` function return `TakerSwapCommand::Finish`, or I can add the command as a mutable parameter to the `check_watcher_payments` function and modify it inside if necessary. What do you think?

Reply:
Thanks for using command matching for fixing this :)